File: py/data_structures/trie.py
""" This module contains code for a class that represents a (simplified)
prefix trie data structure """
from typing import Dict


class Trie:
    """ This class represents a (simple) prefix Trie data structure.

    The trie is an ordered tree data structure used to store strings.
    Anytime you have a dataset of strings and you'd like to query the
    dataset, you can consider using a Trie.

    Where k is the length of the longest word stored in the trie, and n
    is the total number of words stored in the trie.

    Used for autocomplete, spell checking, IP routing, etc.

    Attributes:
        end_sumbol:
            A string that is equivalent to "*" that represents the end symbol
            for a word in the trie

        root:
            Dictionary containing mapping between string keys, with dictionary
            values, representing all the words stored at the root of the trie
    """

    def __init__(self):
        self.end_symbol = "*"
        self.root = {}

    def insert(self, word: str) -> None:
        """ Inserts a word into the Trie.

        Time:
            O(k) best/average/worst

        Space:
            O(k) best/average/worst

        Where k is the length of the word being inserted into
        the trie

        Args:
            word:
                String representing word to insert into the Trie.
        """
        node = self.root
        self._insert_helper(node, word)

    def _insert_helper(self, node: Dict[str, Dict], word: str) -> None:
        for char in word:
            if char not in node:
                node[char] = {}
            node = node[char]

        node[self.end_symbol] = True

    def lookup(self, word: str) -> bool:
        """ This method takes a string as input and returns a boolean
        indicating whether or not it is currently stored in the Trie.

        Time:
            O(k) best/average/worst

        Space:
            O(1) best/average/worst

        Where k is the length of the input string

        Args:
            word:
                String representing word to lookup in the Trie.

        Returns:
            Boolean value representing whether the string is stored in the Trie
        """
        node = self.root
        return self._lookup_helper(node, word)

    def _lookup_helper(self, node: Dict[str, Dict], word: str) -> bool:
        for char in word:
            if char not in node:
                return False
            node = node[char]
        # Word is only in trie if the last node has the end symbol in it
        return self.end_symbol in node

    def starts_with(self, prefix: str) -> bool:
        """ This method recieves a string as an input which is assumed to be
        a prefix pattern, and then returns a boolean indicating whether or not
        the prefix pattern is present in the trie.

        Time:
            O(k) best/average/worst

        Space:
            O(1) best/average/worst

        Where k is the length of the input string

        Input:
            prefix:
                String representing the prefix pattern to look up in the Trie

        Returns:
            Boolean value representing whether the prefix pattern is stored in
            the Trie
        """
        node = self.root
        return self._starts_with_helper(node, prefix)

    def _starts_with_helper(self, node: Dict[str, Dict], prefix: str) -> bool:
        for char in prefix:
            if char not in node:
                return False
            node = node[char]

        return True

    def delete(self, word: str) -> None:
        """ This method recieves a string as input and deletes the string if
        it is currently stored in the Trie.

        Time:
            O(k) best/average/worst

        Space:
            O(1) best/average/worst

        Where k is the length of the input string

        Args:
            word:
                String representing the word to delete in the Trie
        """
        # confirm word is even in the trie before attempting to delete it
        if not self.lookup(word):
            return
        node = self.root
        self._delete_helper(node, word, 0)

    def _delete_helper(self, node: Dict[str, Dict], word: str,
                       idx: int) -> None:
        if idx == len(word):
            node.pop(self.end_symbol)
            return
        new_node = node[word[idx]]
        self._delete_helper(new_node, word, idx + 1)

        # if there are no chars in the new node then we can safely
        # delete it
        if not new_node:
            node.pop(word[idx])

File: ts/src/data structures/trie.ts
class TrieNode {
    children: Record<string, TrieNode>;

    containsEndOfWord: boolean;

    constructor() {
        this.children = {};
        this.containsEndOfWord = false;
    }
}
/**
 *  This class represents a (simple) prefix Trie data structure.
 *  The Trie is an ordered tree data structure used to store strings.
 *  Anytime you have a dataset of strings and you'd like to query the
 *  dataset, you can consider using a Trie.

 * Where k is the length of a word, and n is the total number of words
 * stored in the trie. Used for autocomplete, spell checking, IP routing, etc. 
 */
class Trie {
    root: TrieNode;

    constructor() {
        /** Root node of the tree
         * @param {object}
         */
        this.root = new TrieNode();
    }

    /**
     * Inserts a word into the Trie. 
     * 
     * Time:
     *- O(k) best/avg/worst
        
     * Space:
     *- O(k) best/avg/worst

     * k - length of the word being inserted into the trie 

     * @param {string} word String representing word to insert into the Trie.

     * @returns {undefined}
     */
    insert(word: string): void {
        const node = this.root;
        Trie._insertHelper(node, word);
    }

    static _insertHelper(node: TrieNode, word: string): void {
        let currNode = node;
        for (let i = 0; i < word.length; i += 1) {
            const char = word[i];
            if (!(char in currNode.children)) {
                currNode.children[char] = new TrieNode();
            }
            currNode = currNode.children[char];
        }
        currNode.containsEndOfWord = true;
    }

    /**
     *  This method takes a string as input and returns a boolean indicating 
     *  whether or not it is currently stored in the Trie. 

     * Time:
     *- O(k) best/avg/worst
    
     * Space:
     *- O(1) best/avg/worst

     * k - length of input string 
     * @param {string} word String representing word to lookup in the Trie
     * @returns {boolean} Boolean representing whether the string is stored
     * in the Trie 
     */
    lookup(word: string): boolean {
        const node = this.root;
        return Trie._lookupHelper(word, node);
    }

    static _lookupHelper(word: string, node: TrieNode): boolean {
        let currNode = node;
        for (let i = 0; i < word.length; i += 1) {
            const char = word[i];
            if (!(char in currNode.children)) {
                return false;
            }
            currNode = currNode.children[char];
        }
        return currNode.containsEndOfWord;
    }

    /**  
     * This method recieves a string as an input which is assumed to be a prefix
     *  pattern, and then returns a boolean indicating whether or not the prefix
     *  pattern is present in the trie. 

     * Time:
     *- O(k) best/avg/worst
        
     * Space:
     *- O(1) best/avg/worst
     *
     * k - length of input string 
     * @param {string} prefix String representing the prefix pattern to look up
     * in the Trie
     * @returns {boolean} Boolean representing whether the prefix pattern is
     * stored in the Trie 
     */
    startsWith(prefix: string): boolean {
        const node = this.root;
        return Trie._startsWithHelper(prefix, node);
    }

    static _startsWithHelper(prefix: string, node: TrieNode): boolean {
        let currNode = node;
        for (let i = 0; i < prefix.length; i += 1) {
            const char = prefix[i];
            if (!(char in currNode.children)) {
                return false;
            }
            currNode = currNode.children[char];
        }
        return true;
    }

    /**
     * This method recieves a string as input and deletes the string if
     * it is currently stored in the Trie.
     * 
     * Time:
     *- O(k) best/avg/worst
      
     * Space:
     *- O(1) best/avg/worst 

     * k - length of input string
     * @param {string} word String representing the word to delete in the Trie 
     * @returns {null} 
     */
    delete(word: string): void {
        if (this.lookup(word)) {
            const node = this.root;
            Trie._deleteHelper(node, word, 0);
        }
    }

    static _deleteHelper(node: TrieNode, word: string, idx: number): undefined {
        if (idx === word.length) {
            node.containsEndOfWord = false;
            return;
        }
        const newNode = node.children[word[idx]];
        Trie._deleteHelper(newNode, word, idx + 1);

        if (Object.keys(newNode.children).length === 0) {
            delete node.children[word[idx]];
        }
    }
}

export default Trie;
